File: python/Advanced Graphs/swim_rising_water.py
from typing import List
import heapq

class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        """
        Given an N x N grid where each square grid[r][c] represents the elevation at that point,
        return the least time required to swim from the top left to the bottom right.
        You can swim to 4-directionally adjacent squares if the elevation is at most the current time.
        The time taken is the maximum elevation encountered along the path.

        Approach:
        - Use a variation of Dijkstra's algorithm (min-heap/priority queue) to always expand the lowest elevation cell reachable.
        - At each step, move to the neighbor with the smallest possible elevation/time.
        - The answer is the first time we reach the bottom-right cell.

        Time Complexity: O(N^2 * log N)
            - Each cell is pushed/popped from the heap at most once (N^2 cells)
            - Each heap operation is O(log N^2) = O(log N)
        Space Complexity: O(N^2)
            - For the visited set and the heap
        """
        N = len(grid)
        visit = set()
        minH = [[grid[0][0], 0, 0]] # (time or max-height, r, c)
        directions =[[0, 1], [0, -1], [1, 0], [-1, 0]]
        
        visit.add((0, 0))
        while minH:
            t, r, c = heapq.heappop(minH)

            # If we've reached the bottom-right cell, return the time (max elevation on path)
            if r == N - 1 and c == N - 1:
                return t
            
            # Explore all 4 directions
            for dr, dc in directions:
                neiR, neiC = r + dr, c + dc
                # Check bounds and if already visited
                if (neiR < 0 or neiC < 0 or neiR == N or neiC == N or (neiR, neiC) in visit):
                    continue
                visit.add((neiR, neiC))
                # The time to enter the neighbor is the max of current time and neighbor's elevation
                heapq.heappush(minH, [max(t, grid[neiR][neiC]), neiR, neiC])      
        # If unreachable (should not happen for valid inputs)
        return -1  


def test_swim_in_water():
    solution = Solution()
    # Test case 1: Simple 2x2 grid
    grid1 = [
        [0, 2],
        [1, 3]
    ]
    expected1 = 3
    result1 = solution.swimInWater(grid1)
    print(f"Test 1: {result1 == expected1}, Expected: {expected1}, Got: {result1}")

    # Test case 2: 3x3 grid
    grid2 = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8]
    ]
    expected2 = 8
    result2 = solution.swimInWater(grid2)
    print(f"Test 2: {result2 == expected2}, Expected: {expected2}, Got: {result2}")

    # Test case 3: Path with detour
    grid3 = [
        [0, 2, 1, 3],
        [2, 3, 4, 4],
        [3, 2, 5, 5],
        [6, 7, 8, 9]
    ]
    expected3 = 9  # Corrected expected value
    result3 = solution.swimInWater(grid3)
    print(f"Test 3: {result3 == expected3}, Expected: {expected3}, Got: {result3}")

    # Test case 4: Single cell
    grid4 = [[0]]
    expected4 = 0
    result4 = solution.swimInWater(grid4)
    print(f"Test 4: {result4 == expected4}, Expected: {expected4}, Got: {result4}")

    # Test case 5: 2x2 grid, must wait
    grid5 = [
        [0, 100],
        [1, 2]
    ]
    expected5 = 2  # Corrected expected value
    result5 = solution.swimInWater(grid5)
    print(f"Test 5: {result5 == expected5}, Expected: {expected5}, Got: {result5}")

if __name__ == "__main__":
    test_swim_in_water()  

File: C++/Advanced Graphs/swim_rising_water.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        /*
        Approach:
        - Dijkstra-like BFS on a grid using a min-heap (priority queue).
        - Always move to the reachable cell with the smallest elevation.
        - The result is the max elevation along the path when reaching (N-1, N-1).

        Time Complexity: O(N^2 * log N)
            - N^2 cells in the grid.
            - Each heap operation is O(log N^2) ~ O(log N)

        Space Complexity: O(N^2)
            - Visited grid and the min-heap.
        */

        int N = grid.size();

        vector<vector<bool>> visited(N, vector<bool>(N, false));

        // Min-heap of (time, row, col), where time is the max elevation so far.
        typedef tuple<int, int, int> T;
        priority_queue<T, vector<T>, greater<T>> minH;

        minH.push(make_tuple(grid[0][0], 0, 0));
        visited[0][0] = true;

        vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

        while (!minH.empty()) {
            T top = minH.top();
            minH.pop();

            int time = get<0>(top);
            int r = get<1>(top);
            int c = get<2>(top);

            // If bottom-right is reached, return the time
            if (r == N - 1 && c == N - 1)
                return time;

            for (auto& dir : directions) {
                int nr = r + dir.first;
                int nc = c + dir.second;

                if (nr >= 0 && nc >= 0 && nr < N && nc < N && !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    minH.push(make_tuple(max(time, grid[nr][nc]), nr, nc));
                }
            }
        }

        return -1; // Should not reach here if input is valid
    }
};

/*
================ LINE BY LINE EXPLANATION =====================

#include <bits/stdc++.h>
- Includes all standard C++ libraries.

using namespace std;
- So you don't need to prefix with std::

class Solution
- Defines a class Solution.

int swimInWater(vector<vector<int>>& grid)
- Function that finds the minimum time to swim across the grid.

int N = grid.size();
- Stores grid size.

vector<vector<bool>> visited(N, vector<bool>(N, false));
- Keeps track of which cells are visited.

typedef tuple<int, int, int> T;
- Defines a type alias for a tuple of (time, row, col).

priority_queue<T, vector<T>, greater<T>> minH;
- Min-heap where smallest time is at top.

minH.push(make_tuple(grid[0][0], 0, 0));
- Start with the top-left cell.

visited[0][0] = true;
- Mark start cell as visited.

vector<pair<int, int>> directions
- Defines the 4 movement directions.

while (!minH.empty())
- While there are cells to explore.

T top = minH.top();
- Get cell with smallest elevation seen so far.

minH.pop();
- Remove it from the heap.

int time = get<0>(top), r = get<1>(top), c = get<2>(top);
- Extract time, row, and col from tuple.

if (r == N-1 && c == N-1)
- If we reached bottom-right, return time.

for (auto& dir : directions)
- Check all 4 adjacent cells.

int nr = r + dir.first, nc = c + dir.second;
- Compute neighbor coordinates.

if (...) // check bounds and visited
- Skip invalid or already visited cells.

visited[nr][nc] = true;
- Mark as visited.

minH.push(make_tuple(max(time, grid[nr][nc]), nr, nc));
- Push neighbor with updated max elevation on path.

return -1;
- Fallback if no path found (should not occur).

===============================================================
*/
